### PR TITLE
Optimize tree walk to avoid too depth function call stack.

### PR DESCRIPTION
**Pull request by rubenvp8510**
## Which problem is this PR solving?
- This Resolves https://github.com/jaegertracing/jaeger-ui/issues/320

## Short description of the changes
- After taking a look a the Issue and run the test code mentioned in the issue, it seems like the UI is freezing on the tree walk method when the tree is too depth. After did some tests I think that the "recursive" function calling could be the problem.

I did some tests and it seems to behave better with an iterative walk algorithm. This partially solves the problem, because the render of too many components is slow. We could probably need to impose a limit on the depth (But at least now is not freezing anymore)

### CODE REVIEW

File: packages/jaeger-ui/src/utils/TreeNode.js [hunk 1]
```diff
@@ -83,7 +83,18 @@ export default class TreeNode {
...
 
   walk(fn, depth = 0) {
-    TreeNode.iterFunction(fn, depth)(this);
-    this.children.forEach(child => child.walk(fn, depth + 1));
+    const nodeStack = [];
+    let actualDepth = depth;
+    nodeStack.push({ node: this, depth: actualDepth });
+    const pushChildren = n => nodeStack.push({ node: n, depth: actualDepth });
+    while (nodeStack.length) {
+      const { node, nodeDepth } = nodeStack.pop();
```
Comment (tiffon) [inline packages/jaeger-ui/src/utils/TreeNode.js]: I think this is destructing the property name:

```js
const { node, depth: nodeDepth } = nodeStack.pop();
```

File: packages/jaeger-ui/src/utils/TreeNode.js [hunk 2]
```diff
@@ -83,7 +83,18 @@ export default class TreeNode {
...
+      TreeNode.iterFunction(fn, nodeDepth)(node);
+      actualDepth++;
```
[the remark this answers is not in the document]
Comment (rubenvp8510) [reply packages/jaeger-ui/src/utils/TreeNode.js]: I need to review this case, it seems like you're right and I need to verify the logic behind the deep computation.

File: packages/jaeger-ui/src/utils/TreeNode.js (hunk 1 above)
Comment (rubenvp8510) [reply packages/jaeger-ui/src/utils/TreeNode.js]: I like it :)

File: packages/jaeger-ui/src/utils/TreeNode.js [hunk 3]
```diff
@@ -83,7 +83,18 @@ export default class TreeNode {
...
   walk(fn, depth = 0) {
-    TreeNode.iterFunction(fn, depth)(this);
-    this.children.forEach(child => child.walk(fn, depth + 1));
+    const nodeStack = [];
+    let actualDepth = depth;
+    nodeStack.push({ node: this, depth: actualDepth });
+    const pushChildren = n => nodeStack.push({ node: n, depth: actualDepth });
+    while (nodeStack.length) {
+      const { node, nodeDepth } = nodeStack.pop();
+      TreeNode.iterFunction(fn, nodeDepth)(node);
```
[the remark this answers is not in the document]
Comment (rubenvp8510) [reply packages/jaeger-ui/src/utils/TreeNode.js]: Absolutely!

File: packages/jaeger-ui/src/utils/TreeNode.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rubenvp8510) [reply packages/jaeger-ui/src/utils/TreeNode.js]: Yes,  the way I'm actually using for depth calculation is wrong, I'm gonna fix it, add a test will and update this PR.

[the remark this answers is not in the document]
Comment (rubenvp8510) [reply packages/jaeger-ui/src/utils/TreeNode.js]: I've already fixed this and written one test.

**Comment by rubenvp8510** [comment]
Ready for another review :)

**Comment by tiffon** [comment]
### Awesome work! Thanks so much for making these changes! :tada:

**Pull request merged**